File: src/utils/http.ts
/**
 * Configuration options for the HTTP client.
 */
export interface HttpClientOptions {
  /** Base URL for API requests */
  baseUrl: string;
  /** Optional API key for authenticated requests */
  apiKey?: string;
  /** Optional additional headers to include in all requests */
  headers?: Record<string, string>;
}

/**
 * Custom error class for DFlow API errors.
 *
 * Thrown when the API returns a non-2xx status code.
 * Contains the HTTP status code and response body for debugging.
 *
 * @example
 * ```typescript
 * import { DFlowApiError } from 'dflow-sdk';
 *
 * try {
 *   const market = await dflow.markets.getMarket('invalid-ticker');
 * } catch (error) {
 *   if (error instanceof DFlowApiError) {
 *     console.error(`API Error ${error.statusCode}: ${error.message}`);
 *     console.error('Response:', error.response);
 *   }
 * }
 * ```
 */
export class DFlowApiError extends Error {
  /**
   * Create a new API error.
   *
   * @param message - Error message
   * @param statusCode - HTTP status code from the response
   * @param response - Parsed response body (if available)
   */
  constructor(
    message: string,
    public statusCode: number,
    public response?: unknown
  ) {
    super(message);
    this.name = 'DFlowApiError';
  }
}

/**
 * Internal HTTP client for making API requests.
 *
 * Handles request construction, authentication headers, and response parsing.
 * Used internally by all API classes.
 */
export class HttpClient {
  private baseUrl: string;
  private apiKey?: string;
  private defaultHeaders: Record<string, string>;

  /**
   * Create a new HTTP client.
   *
   * @param options - Client configuration
   */
  constructor(options: HttpClientOptions) {
    // Ensure baseUrl ends with /
    this.baseUrl = options.baseUrl.endsWith('/') ? options.baseUrl : options.baseUrl + '/';
    this.apiKey = options.apiKey;
    this.defaultHeaders = {
      'Content-Type': 'application/json',
      Accept: 'application/json',
      ...options.headers,
    };
  }

  private getHeaders(): Record<string, string> {
    const headers = { ...this.defaultHeaders };
    if (this.apiKey) {
      headers['x-api-key'] = this.apiKey;
    }
    return headers;
  }

  private buildUrl(path: string, params?: object): string {
    // Remove leading slash from path to avoid URL resolution issues
    const cleanPath = path.startsWith('/') ? path.slice(1) : path;
    const url = new URL(cleanPath, this.baseUrl);

    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          url.searchParams.append(key, String(value));
        }
      });
    }

    return url.toString();
  }

  /**
   * Make a GET request.
   *
   * @param path - API endpoint path
   * @param params - Optional query parameters
   * @returns Parsed JSON response
   * @throws {@link DFlowApiError} if the request fails
   */
  async get<T>(path: string, params?: object): Promise<T> {
    const url = this.buildUrl(path, params);

    const response = await fetch(url, {
      method: 'GET',
      headers: this.getHeaders(),
    });

    return this.handleResponse<T>(response);
  }

  /**
   * Make a POST request.
   *
   * @param path - API endpoint path
   * @param body - Optional request body (will be JSON serialized)
   * @returns Parsed JSON response
   * @throws {@link DFlowApiError} if the request fails
   */
  async post<T>(path: string, body?: unknown): Promise<T> {
    const url = this.buildUrl(path);

    const response = await fetch(url, {
      method: 'POST',
      headers: this.getHeaders(),
      body: body ? JSON.stringify(body) : undefined,
    });

    return this.handleResponse<T>(response);
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    const text = await response.text();

    if (!response.ok) {
      let errorBody: unknown;
      try {
        errorBody = JSON.parse(text);
      } catch {
        errorBody = text;
      }

      throw new DFlowApiError(
        `HTTP ${response.status}: ${response.statusText}`,
        response.status,
        errorBody
      );
    }

    try {
      return JSON.parse(text) as T;
    } catch {
      throw new DFlowApiError(
        `Failed to parse response as JSON`,
        response.status,
        text
      );
    }
  }

  /**
   * Update the API key for subsequent requests.
   *
   * @param apiKey - New API key to use
   */
  setApiKey(apiKey: string): void {
    this.apiKey = apiKey;
  }
}

File: packages/python/src/dflow/utils/http.py
"""HTTP client for DFlow API requests."""

from typing import Any

import httpx


class DFlowApiError(Exception):
    """Custom error class for DFlow API errors.

    Thrown when the API returns a non-2xx status code.
    Contains the HTTP status code and response body for debugging.

    Example:
        >>> try:
        ...     market = dflow.markets.get_market("invalid-ticker")
        ... except DFlowApiError as e:
        ...     print(f"API Error {e.status_code}: {e}")
        ...     print(f"Response: {e.response}")
    """

    def __init__(self, message: str, status_code: int, response: Any = None):
        """Create a new API error.

        Args:
            message: Error message
            status_code: HTTP status code from the response
            response: Parsed response body (if available)
        """
        super().__init__(message)
        self.status_code = status_code
        self.response = response


class HttpClient:
    """Internal HTTP client for making API requests.

    Handles request construction, authentication headers, and response parsing.
    Used internally by all API classes.
    """

    def __init__(
        self,
        base_url: str,
        api_key: str | None = None,
        headers: dict[str, str] | None = None,
        timeout: float = 30.0,
    ):
        """Create a new HTTP client.

        Args:
            base_url: Base URL for API requests
            api_key: Optional API key for authenticated requests
            headers: Optional additional headers to include in all requests
            timeout: Request timeout in seconds (default: 30.0)
        """
        # Ensure base_url ends with /
        self.base_url = base_url.rstrip("/") + "/"
        self.api_key = api_key
        self._default_headers = headers or {}
        self._client = httpx.Client(
            base_url=self.base_url,
            headers=self._build_headers(),
            timeout=timeout,
        )

    def _build_headers(self) -> dict[str, str]:
        """Build request headers including auth if available."""
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            **self._default_headers,
        }
        if self.api_key:
            headers["x-api-key"] = self.api_key
        return headers

    def get(self, path: str, params: dict[str, Any] | None = None) -> Any:
        """Make a GET request.

        Args:
            path: API endpoint path
            params: Optional query parameters

        Returns:
            Parsed JSON response

        Raises:
            DFlowApiError: If the request fails
        """
        # Clean path and filter None params
        clean_path = path.lstrip("/")
        clean_params = (
            {k: v for k, v in params.items() if v is not None} if params else None
        )

        response = self._client.get(clean_path, params=clean_params)
        return self._handle_response(response)

    def post(self, path: str, json: Any = None) -> Any:
        """Make a POST request.

        Args:
            path: API endpoint path
            json: Optional request body (will be JSON serialized)

        Returns:
            Parsed JSON response

        Raises:
            DFlowApiError: If the request fails
        """
        clean_path = path.lstrip("/")
        response = self._client.post(clean_path, json=json)
        return self._handle_response(response)

    def _handle_response(self, response: httpx.Response) -> Any:
        """Handle API response, raising errors for non-2xx status codes."""
        if not response.is_success:
            try:
                error_body = response.json()
            except Exception:
                error_body = response.text

            raise DFlowApiError(
                f"HTTP {response.status_code}: {response.reason_phrase}",
                response.status_code,
                error_body,
            )

        try:
            return response.json()
        except Exception:
            raise DFlowApiError(
                "Failed to parse response as JSON",
                response.status_code,
                response.text,
            )

    def set_api_key(self, api_key: str) -> None:
        """Update the API key for subsequent requests.

        Args:
            api_key: New API key to use
        """
        self.api_key = api_key
        # Rebuild client with new headers
        self._client = httpx.Client(
            base_url=self.base_url,
            headers=self._build_headers(),
            timeout=self._client.timeout,
        )

    def close(self) -> None:
        """Close the HTTP client."""
        self._client.close()

    def __enter__(self) -> "HttpClient":
        return self

    def __exit__(self, *args: Any) -> None:
        self.close()
